### ISSUE TITLE
aspect='equal' ignored when parasitic axis added with twinx/twiny

### ISSUE THREAD

**Issue by ndeadly** [matplotlib/matplotlib#7654]
Consider an axis with equal aspect ratio

```
from matplotlib import pyplot as plt

fig = plt.figure()
ax1 = fig.add_subplot(111, aspect='equal')
ax1.grid()
plt.show()
```
![figure_1-5](https://cloud.githubusercontent.com/assets/24677491/21387318/60fa2f44-c777-11e6-9421-140bbecf799d.png)

---

Adding a parasitic axis with `twinx()` causes the plot to completely ignore `aspect='equal'`

```
from matplotlib import pyplot as plt

fig = plt.figure()
ax1 = fig.add_subplot(111, aspect='equal')
ax1.grid()
ax2 = ax1.twinx()
plt.show()
```

![figure_1-6](https://cloud.githubusercontent.com/assets/24677491/21387344/7a8d06ca-c777-11e6-9339-882f7e9d847b.png)

---

Creating the primary axis with the keyword argument `adjustable='box-forced'` causes it to obey `aspect='equal'` but the parasitic axis does not.

```
from matplotlib import pyplot as plt

fig = plt.figure()
ax1 = fig.add_subplot(111, adjustable='box-forced', aspect='equal')
ax1.grid()
ax2 = ax1.twinx()
plt.show()
```

![figure_1-7](https://cloud.githubusercontent.com/assets/24677491/21387390/bee23eb2-c777-11e6-9e0f-86e4b07b4bf8.png)

---

Interestingly, everything works if the primary axis is created with `host_subplot` (note `adjustable='box-forced'` is still required). However, it seems to be prone to a similar issue to #7648 if `figure.autolayout` is set

```
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1 import host_subplot

fig = plt.figure()
ax1 = host_subplot(111, adjustable='box-forced', aspect='equal')
ax1.grid()
ax2 = ax1.twinx()
plt.show()
```
![figure_1-8](https://cloud.githubusercontent.com/assets/24677491/21387757/7ddc4faa-c779-11e6-9f80-03edb2c76d7d.png)

Platform: Windows 7
matplotlib: 1.5.3
python: 2.7.12

**Comment by efiring** [comment]
Two comments related to this:
1) 'box-forced' is inherently designed to break things.  I have always maintained it was a bad idea.  The best that can be said for it is that inclusion of the word 'forced' should warn the user to expect breakage.  Some things will work, others won't.  You are dealing with an over-determined system, so things will work only if all constraints turn out to be consistent.
2) Your statement that "Adding a parasitic axis with twinx() causes the plot to completely ignore aspect='equal'" is not correct.  The original axis still has a *data* aspect ratio of 1, which it is achieving by adjusting the data limits. `twinx` turns on axis sharing, which incompatible with aspect ratio control unless the adjustable is 'datalim'.

We used to have a warning when the adjustable parameter was switched from 'box' to 'datalim' in this case. It is presently commented out.  I think we need to uncomment it, and maybe even give it a longer explanation.

**Comment by efiring** [comment]
This is partly addressed by #10033, but it looks like more can be done.  Instead of setting adjustable to 'datalim' on twinned axes, I think that it would be possible to check at drawing time whether the aspect ratio is specified for more than a single axes in a twinned pair, and switch to datalim only in that case.  A 'box' adjustable should work for twinned axes with only one aspect ratio specified; or perhaps also if specified aspect ratios are identical.

**Comment by DhrumilSoft** [comment]
I was investigating a fix for this bug and even after the partial changes made by 10033 there still seem to be issues and I need clarification on how it's supposed to behave. 

1. What should the adjustable be when an axes and its twin both have aspect set to 'auto'? 

2. What if the aspects for both axes are the same? Should the adjustable be 'box' in these cases?

3. What should happen in cases where the aspects are different like 'auto' and 'equal'?  I tried to set this to datalim in the drawing time but I noticed that the current bug still persists with one of the axis going to 0.8 and the other one going to 1.0. Is this expected?
